include/dlaf/multiplication/general.h

```diff
@@ -64,6 +64,39 @@ void generalMatrix(const blas::Op opA, const blas::Op opB, const T alpha, Matrix
     DLAF_UNIMPLEMENTED(opA, opB);
 }
 
+/// General sub-matrix distributed multiplication, computing
+/// C = alpha * A * B + beta * C
+///
+/// @param  mat_a contains the input matrix A.
+/// @pre @p mat_a is distributed according to @p grid
+///
+/// @param  mat_b contains the input matrix B.
+/// @pre @p mat_b is distributed according to @p grid
+///
+/// @param  mat_c On entry it contains the input matrix C. On exit matrix tiles in the range will be
+///         overwritten with the result, while others are left untouched.
+/// @pre @p mat_c is distributed according to @p grid
+///
+/// @pre multipliable_sizes(mat_a.size(), mat_b.size(), mat_c.size(), opA, opB)
+/// @pre multipliable_sizes(mat_a.tile_size(), mat_b.tile_size(), mat_c.tile_size(), opA, opB)
+/// @pre multipliable_sizes(mat_a.tile_size_of({0, 0}), mat_b.tile_size_of({0, 0}),
+///      mat_c.tile_size_of({0, 0}), opA, opB)
+template <Backend B, Device D, class T>
+void generalMatrix([[maybe_unused]] comm::CommunicatorGrid grid,
+                   common::Pipeline<comm::Communicator>& row_task_chain,
+                   common::Pipeline<comm::Communicator>& col_task_chain, const T alpha,
+                   MatrixRef<const T, D>& mat_a, MatrixRef<const T, D>& mat_b, const T beta,
+                   MatrixRef<T, D>& mat_c) {
+  DLAF_ASSERT(matrix::equal_process_grid(mat_a, grid), mat_a, grid);
+  DLAF_ASSERT(matrix::equal_process_grid(mat_b, grid), mat_a, grid);
+  DLAF_ASSERT(matrix::equal_process_grid(mat_c, grid), mat_a, grid);
```

Comment:
Why are these not simply:
```suggestion
  DLAF_ASSERT(matrix::same_process_grid(mat_a, mat_b), mat_a, mat_b);
  DLAF_ASSERT(matrix::same_process_grid(mat_b, mat_c), mat_b, mat_c);
```
? I guess the `grid` kind of additionally acts as a tag here to say that this is a distributed matrix multiplication? But then the pipelines already kind of do that as well...

Reply:
Good point! Just updated in https://github.com/eth-cscs/DLA-Future/pull/1022/commits/4f9514d1f3e2da25ba11729f8d3e09e5524e2090

Moreover, if I'm not mistaken, the requirement could be even slightly more relaxed (at least for the current implementation of the case we support).

@rasolca what do you think?

Reply:
About the slightly more relaxed requirement, I was wrong.

About the check, as soon as we will get #993 we can add the check about the "matching" between matrices grid and pipelines.

For what concerns this PR, I think this change is enough 😉

Reply:
Should I update something in #993 now that this is merged or would you do the change you're talking about separately?